### ISSUE TITLE
Poor signal-to-noise ratio in nix build log

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nix#443]
Carrying over from https://github.com/NixOS/nixpkgs/issues/5723:

> This all falls under the nix usability umbrella, but it would be nice if by default most packages hid command output unless something went wrong, only giving me high-level feedback unless I asked for more verbosity. The log output currently makes it almost impossible to figure out what's going on unless you use search functionality to skip over all the noise.
> 
> I imagine fixing this would amount to wrapping the calls to the various *phase values in buffering wrappers that save their subprocess output but discard it if there are no errors (or write it to a log file). Not sure if some people like seeing all that output, but it doesn't seem like a great default.

**Comment by shlevy** [comment]
IMO a better option is to fix this at the nix level (presumably that's why you opened this here?). One simple option is just to have a mode where nix doesn't print anything unless a build fails, at which point it prints just some subset of the failed build log. Alternatively we could add some more complex interface for the build to signal that something should be shown to the user

**Comment by copumpkin** [comment]
That's why it's here, yeah :smile:. I think the ideal would be what you say, potentially with some directives (color codes!!1!!!1) to allow us to signal interesting things in build logs (for those you do look at). So the different phases, perhaps, and perhaps nesting.

**Comment by copumpkin** [comment]
Also, could it be possible to do something with a pty to get color output from our subprocesses? Then a separate process could scrub that from logfiles and the like.

**Comment by iElectric** [comment]
Did you see https://github.com/qknight/nix-build-view

**Comment by copumpkin** [comment]
I hadn't seen that! It looks very promising. It scares me how detached it seems from the main codebase though. Is the plan to somehow merge it back in when it's done? I want efficient presentation of information by default, ideally :smile:

**Comment by wmertens** [comment]
What would also be nice is that nix-build, instead of adding escape codes, nix adds tags like `{e}` in front of stderr output and `{n}` in front of lines added by Nix, as well as escaping regular output with `{o}` if it starts with `{`. Something like that anyway.
Then log files are human readable and all the signaling is out-of-band. When output should be printed, it can be colorized as desired.

**Comment by vcunat** [comment]
I use ``-Q`` for almost all nix building invocations, and then potentially inspect the build logs separately (typically just failed ones). The verbose default seems mostly useless, due to the sheer volume and unrelated parallel builds together, so I believe we should **change the default**.

As a compromise between ``-Q`` and the fully verbose way, we might default to @shlevy's suggestion: output the last ~20 lines iff a build fails (avoiding intermixing multiple build logs).

I'd wait for some opinion from @edolstra before proceeding (as he's authored almost all code in this repo).

**Comment by wmertens** [comment]
See also
https://nixos.org/wiki/GSOC_2015_ideas_list#Improve_nix-build_output_for_post-processing

**Comment by ip1981** [comment]
`nix-build ... -Q ... | whatupnix`

```bash
#!/usr/bin/env bash

GAWK=${GAWK:-gawk}
NIX_STORE=${NIX_STORE:-nix-store}

broken=$(mktemp)
trap 'rm -f -- "$broken"' EXIT

$GAWK -v broken="$broken" -f <(cat - <<- 'AWK'
  match($0, /builder for .*(\/nix\/store\/.+\.drv).* failed/, m) {
    print m[1] >> broken
  }
  { print $0 }
AWK
)

while read -r drv; do
  echo "=*= $drv build log =*=" >&2

  # FIXME: nix-store tends to use pager:
  $NIX_STORE -l "$drv" | cat >&2

  echo >&2
done < "$broken"

exit 0
```

**Comment by unode** [comment]
Risking that I'm asking in an inappropriate place...

Would it be a reasonable addition to include a line echoing Start/End of any given build phase?
```
echo "Starting preCheck phase"
preCheck
echo "End preCheck phase"
...
# In case of failures
echo "Build failed during preCheck phase ..."
```
I often find myself guessing the phase to target when things go wrong.

**Comment by edolstra** [comment]
Closed via the Nix 2.0 progress indicator.
